File: tagging_tools/ELFSectionTagger.py
from elftools.elf.elffile import ELFFile
from elftools.elf.constants import SH_FLAGS

# really not needed, but it's where you find RangeFile
import TaggingUtils

RWX_X = 'elf.Section.SHF_EXECINSTR'
RWX_R = 'elf.Section.SHF_ALLOC'
RWX_W = 'elf.Section.SHF_WRITE'

def generate_rwx_ranges(ef, range_file):
    sections = list(ef.iter_sections())
    for i,s in enumerate(sections):
          flags = s['sh_flags']
          start = s['sh_addr']
          end = start + s['sh_size']
          if (end % 4) != 0:
              end += 4 - (end % 4)
          if flags & SH_FLAGS.SHF_EXECINSTR:
               range_file.write_range(start, end, RWX_X)
               range_file.write_range(start, end, RWX_R)
               if (".init" in s.name or
                  ".exit" in s.name):
                   if i != len(sections)-1:
                       end = sections[i + 1]['sh_addr']
                   print('WX {0}: 0x{1:X} - 0x{2:X}'.format(s.name, start, end))
                   range_file.write_range(start, end, RWX_W)
               else:
                   print('X {0}: 0x{1:X} - 0x{2:X}'.format(s.name, start, end))
          elif flags & SH_FLAGS.SHF_WRITE:
               range_file.write_range(start, end, RWX_W)
               print('W {0}: 0x{1:X} - 0x{2:X}'.format(s.name, start, end))
          elif flags & SH_FLAGS.SHF_ALLOC:
               range_file.write_range(start, end, RWX_R)
               print('R {0}: 0x{1:X} - 0x{2:X}'.format(s.name, start, end))

File: tagging_tools/elf_section_tagger.cc
#include <cstdint>
#include <gelf.h>
#include <iostream>
#include <string>
#include "elf_loader.h"
#include "range_map.h"
#include "reporter.h"

namespace policy_engine {

static const std::string RWX_X = "elf.Section.SHF_EXECINSTR";
static const std::string RWX_R = "elf.Section.SHF_ALLOC";
static const std::string RWX_W = "elf.Section.SHF_WRITE";

void add_rwx_ranges(range_map_t& range_map, const elf_image_t& ef, reporter_t& err) {
  for (const elf_section_t& section : ef.sections) {
    if (section.flags & SHF_EXECINSTR) {
      range_map.add_range(section.address, section.end_address(), {RWX_R, RWX_X});
      err.info("X %s: %#lx - %#lx\n", section.name, section.address, section.end_address());
    } else if (section.flags & SHF_WRITE) {
      range_map.add_range(section.address, section.end_address(), RWX_W);
      err.info("W %s: %#lx - %#lx\n", section.name, section.address, section.end_address());
    } else if (section.flags & SHF_ALLOC) {
      range_map.add_range(section.address, section.end_address(), RWX_R);
      err.info("R %s: %#lx - %#lx\n", section.name, section.address, section.end_address());
    }
  }
}

}